Added remove-ident function with specs to primitives

currentoor opened the pull request:
@awkay you said

> 1. Helper functions for targeted state manipulation when you know what you want to remove. Not really GC, but helpful. These would be functions like remove-ident, delete-form-state, etc.
> 2. Semi-automated GC...e.g. (recursive-delete some-ident), which delete the table entry at ident, and ALSO follows idents starting at that entity, and removes all children. This requires something like a white-list/black-list of table names, or a set of "isComponent" Datomic-like indicators to prevent unintentional traversal.
> 3. Fully-automated GC. The function scans entire app state removing things that are not referenced. This one also needs "help" in that unions may not be currently pointing at data that needs to be kept.

At the moment I only needed (1), so that's all I did. If the need for (2) or (3) arrises, I'll submit another PR.

Also, after this is merged I can update the book examples to use it. 

https://github.com/fulcrologic/fulcro/issues/216

awkay commented:
So, the API of remove-ident should probably include a flag to indicate if the targeted table entry was "owned" and should be cleaned up.  Probably a multi-arity version:

`(remove-ident state-map ident path true)` means remove the ident from the vector at path, and *also* `(update state-map (first ident) dissoc (second ident))`, whereas a missing 4th arg would mean the table entry stays.  Thoughts?

currentoor commented:
I was initially thinking this would be a low-level function only concerned with removing an ident from a path.

I usually have the logic for *removing ident from list* in one mutation and the logic for *removing the entity* in another mutation. The former is located near the UI component that needs that list of idents.

I prefer keeping them separate. Feels cleaner and more explicit. 

But if we're going to combine them, we should probably call it something else no? Or make a function called `remove-entity` which removes the entity from its table, and takes a list of paths (so calls `remove-ident` on each of those paths). What do you think?

 Also I forgot to add the form state cleanup function to this PR. I'll do that now.

wilkerlucio commented:
I think we should use fulcro incubator for those kinds of helpers, it's hard to figure how good they are until we put then for a trial period, and I think fulcro incubator is the right place to let these functions grow. This could be added to https://github.com/fulcrologic/fulcro-incubator/blob/develop/src/fulcro/incubator/db_helpers.cljc

currentoor commented:
I agree with @wilkerlucio for the higher level functions, however the lower-level ones in this PR IMHO obviously need to be in Fulcro.

The logic of `remove-ident` is already duplicated 3 times in the book. The logic in `delete-form-state*` requires you to first figure out how fulcro manages form state copies, we don't require people to figure that out when they want to create a form-state copy, so why make them figure it out when they want to remove a form-state copy.

wilkerlucio commented:
Makes sense, I don't have a strong opinion on this one in specific, not against having it on primitives, thanks for clarifying @currentoor !

awkay commented on src/main/fulcro/ui/form_state.cljc:
```diff
@@ -546,6 +546,20 @@
   :args (s/cat :state map? :entity-ident util/ident? :field (s/? keyword?))
   :ret map?)
 
+(defn delete-form-state*
```
I think remove the varargs, and make it detect a vector of idents or a single ident. That way threading will still work well with it (since you cannot easily use apply in threading).

awkay commented on src/main/fulcro/client/primitives.cljc:
```diff
@@ -3179,6 +3179,15 @@
                   (throw (ex-info "Unknown post-op to merge-state!: " {:command command :arg data-path})))))
       state actions)))
 
+(defn remove-ident
```
Move to mutations, and rename `remove-ident*`

awkay commented on src/main/fulcro/client/primitives.cljc:
```diff
@@ -3228,6 +3237,14 @@
     "The state has to be an atom. Use 'integrate-ident' instead.")
   (apply swap! state integrate-ident ident named-parameters))
 
+(defn remove-ident!
```
Move to mutations, and make the signature:

`(remove-ident! this field ident)`

so it can be used in a component that owns a list instead of writing a transact.

awkay commented:
While we're at it, please copy `integrate-ident` to `mutations/integrate-ident*`, alias the old one to the new one, make the old doc string say DEPRECATED, SEE (the other one).  Also add deprecation to `integrate-ident!`and say to combine `integrate-ident*` into your mutations instead.

Add log warnings about deprecation that tell them how to port.

awkay commented:
Also, be sure to add to CHANGELOG, and DevelopersGuide.adoc

currentoor commented:
@awkay I made the edits you suggested and updated the CHANGELOG. Also, I updated the dev guide in one place. I could have updated more, but it looks you were trying to make point about how simple updating a normalized app-state is. I felt like using the mutation helper `remove-ident*` would obfuscate that point. Please let me know if you disagree.

awkay commented on src/book/book/demos/parent_child_ownership_relations.cljs:
```diff
@@ -23,7 +23,7 @@
         (fn [s]
```
Why not delete filter-item?

awkay commented on src/main/fulcro/client/primitives.cljc:
```diff
@@ -3138,46 +3138,12 @@
-                             (assoc-in state data-path ident))
-                  (throw (ex-info "Unknown post-op to merge-state!: " {:command command :arg data-path})))))
-      state actions)))
+  (log/warn "This function has been been deprecated and will be removed in the future."
```
No one is going to know what "This function" is. Well, they can infer it, but the message should say `integrate-ident is deprecated`.

awkay commented on src/main/fulcro/client/primitives.cljc:
```diff
@@ -3211,22 +3177,13 @@
-  - replace: A vector (path) to a specific location in app-state where this object's ident should be placed. Can target a to-one or to-many.
-   If the target is a vector element then that element must already exist in the vector.
-  "
+  "DEPRECATED: Use fulcro.client.mutations/integrate-ident* in your mutations instead."
```
Same comment: "integrate-ident! is deprecated and will be removed in a future version. Please use..."

awkay commented on src/main/fulcro/ui/form_state.cljc:
```diff
@@ -546,6 +546,24 @@
+              (apply dissoc s ks)))))
+
+(s/fdef delete-form-state*
+        :args (s/cat :state-map map? :entity-idents (s/* util/ident?))
```
I don't think this spec is right. Should be an OR or ident or every ident

awkay commented:
Sorry I'm slow. Lot's going on.  Couple more comments.

currentoor commented:
@awkay I addressed your comments, added `sente-options` to client side websockets, and made the default sente socket type auto (which prefers websockets if available).

Also I tested the websockets changes locally in my app.

awkay commented on src/main/fulcro/websockets.cljs:
```diff
@@ -81,10 +86,12 @@
   - `global-error-callback` - A function (fn [resp] ...) that is called when returned status code from the server is not 200.
   - `auto-retry?` - A boolean (default false). If set to true any network disconnects will lead to infinite retries until
+  - `sente-options` - A map of options that is passed directly to the sente websocket channel construction (see sente docs).
   the network returns. All remote mutations should be idempotent.
```
oops. This line goes with auto-retry

awkay commented on src/main/fulcro/websockets.cljs:
```diff
@@ -97,4 +104,5 @@
                      :stop                  (atom nil)
-                     :req-params            req-params})))
+                     :req-params            req-params
+                     :sente-options         sente-options})))
```
What happens if you don't pass options?  You'd be passing nil to sente...is that ok?

awkay replied:
Also you said something about "auto", but I don't see it.

awkay commented:
Getting closer...just one code thing to test (no sente options) and a docstring fix.

awkay commented on src/main/fulcro/websockets.cljs:
```diff
@@ -97,4 +104,5 @@
                      :stop                  (atom nil)
-                     :req-params            req-params})))
+                     :req-params            req-params
+                     :sente-options         sente-options})))
```
Nevermind!  I missed the call to merge and thought it was a separate option...

awkay commented:
Actually, I misread it. Only that docstring needs a fix, and I can do that.
